File: 29-04-2025/merge-sort.c
function mergeSort(array)
  if length(array) <= 1
    return array

  mid = length(array) / 2
  leftArray = mergeSort(first half of array)
  rightArray = mergeSort(second half of array)

  return merge(leftArray, rightArray)

function merge(leftArray, rightArray)
  resultArray = []
  i = 0
  j = 0

  while i < length(leftArray) and j < length(rightArray)
    if leftArray[i] < rightArray[j]
      append leftArray[i] to resultArray
      i = i + 1
    else
      append rightArray[j] to resultArray
      j = j + 1

  while i < length(leftArray)
    append leftArray[i] to resultArray
    i = i + 1

  while j < length(rightArray)
    append rightArray[j] to resultArray
    j = j + 1

  return resultArray

File: 29-04-2025/merge-sort.py
def mergeSort(array):
    if len(array) <= 1:
        return array
    
    mid = len(array) // 2
    beg = 0
    end = len(array)
    leftArray = mergeSort(array[0:mid])
    rightArray = mergeSort(array[mid:end])

    return merge(leftArray, rightArray)

def merge(leftArray, rightArray):
    resultArray = []
    i = 0
    j = 0

    while i < len(leftArray) and j < len(rightArray):
        if leftArray[i] < rightArray[j]:
            resultArray.append(leftArray[i])
            i += 1
        else:
            resultArray.append(rightArray[j])
            j += 1

    while i < len(leftArray):
        resultArray.append(leftArray[i])
        i += 1
    
    while j < len(rightArray):
        resultArray.append(rightArray[j])
        j += 1

    return resultArray

def main():
    array = [5, 3, 1, 2, 6]
    print(mergeSort(array))

main()